### PR TITLE
[#125] - Enhancing clustering functionality

### PR DESCRIPTION
**Pull request by sunggun-yu**
Transfering rabbitmq-cluster cookbook into rabbitmq cookbook
* Changing attribute prefix : [:rabbitmq-cluster] to [:rabbitmq][:clustering]
* Enhancement on regex : user ruby matcher
* Removing unnecessary resource : node_type on change_cluster_node_type action
* Some other enhancement on provider

Enhancing from rabbitmq-cluster as the manual clustering
* Removing master/slave concept
* Removing cluster node type
* Introducing new attribute that defines all the cluster nodes and its cluster node type : `[:rabbitmq][:clustering][:cluster_nodes]`
* Adding set_cluster_name action

Applying new features and attributes into auto clustring
* Using `default[:rabbitmq][:clustering][:cluster_nodes]` attribute instead of `['rabbitmq']['cluster_disk_nodes']`
* Applying set_cluster_name and change_cluster_node_type for auto clustering
* Long story short, All the nodes will be set as `disc` node in the rabbitmq.config even though you have some ram nodes. however, change_cluster_node_type action will fix the node type as you expected.

Additional changes on the others
* Fixing typo in default receipe for stopping service to set erlang cookie

Attributes example
```
default[:rabbitmq][:clustering][:use_auto_clustering] = false
default[:rabbitmq][:clustering][:cluster_name] = 'seoul_tokyo_newyork'
default[:rabbitmq][:clustering][:cluster_nodes] = [
    {
        :name => 'rabbit@rabbit1',
        :type => 'disc'
    },
    {
        :name => 'rabbit@rabbit2',
        :type => 'ram'
    },
    {
        :name => 'rabbit@rabbit3',
        :type => 'disc'
    }
]
```

### CODE REVIEW

**Comment by cmluciano** [comment]
Thank you for taking the time to work on this. I am taking a look through it now.

Can you fix the style issues that travis is complaining about?

Also I think we might want to a section to the README describing how to use these changes.

**Comment by sunggun-yu** [comment]
@cmluciano 
Sure, I will.

**Comment by sunggun-yu** [comment]
@cmluciano I fixed style issue and rubocop issue. and Travis has passed. please review it.
Thank you.

File: providers/cluster.rb [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+    Chef::Application.fatal!("[rabbitmq_cluster] #{cmd.stderr}")
+  end
+end
+
+# Get cluster status result
+def cluster_status
+  # execute > rabbitmqctl cluster_status | sed "1d" | tr "\n" " " | tr -d " "
+  # rabbitmqctl cluster_status returns "Cluster status of node rabbit@rabbit1 ..." at the first line.
+  # To parse the result string, it is removed by sed "1d"
+  cmd = 'rabbitmqctl cluster_status | sed "1d" | tr "\n" " " | tr -d " "'
```
Comment (cmluciano) [inline providers/cluster.rb]: I'm a little confused by the need to pipe to sed, tr. etc. Is this to try to normalize the output?

File: providers/cluster.rb [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+  cmd = get_shellout(cmd)
+  cmd.run_command
+  begin
+    cmd.error!
+    Chef::Log.info("[rabbitmq_cluster] #{cmd.stdout}")
+  rescue
+    err = cmd.stderr
+    Chef::Log.warn("[rabbitmq_cluster] #{err}")
+    if err.include?('{ok,already_member}')
+      Chef::Log.info('[rabbitmq_cluster] Node is already member of cluster, error will be ignored.')
```
Comment (cmluciano) [inline providers/cluster.rb]: Can you change this to 
```
[rabbitmq_cluster] Node is already a member of the cluster, error will be ignored.
```

File: providers/cluster.rb (hunk 1 above)
Comment (jjasghar) [inline providers/cluster.rb]: Yeah i have to parrot @cmluciano here, this `cmd` scares me. Is there a way we could leverage some ruby or something other then `sed` `tr` `tr`?

Comment (sunggun-yu) [inline providers/cluster.rb]: Yeah, that is for normalize the output. because, rabbitmqctl output, mostly indent and line feed, was different from servers and versions.

File: providers/cluster.rb (hunk 2 above)
Comment (sunggun-yu) [inline providers/cluster.rb]: sure, i'll change. Thank you for the correction ;-)

File: providers/cluster.rb (hunk 1 above)
Comment (sunggun-yu) [inline providers/cluster.rb]: I'l try with ruby.

Comment (sunggun-yu) [inline providers/cluster.rb]: @jjasghar @cmluciano 

It would be changed like this.
```
def cluster_status
  # execute > rabbitmqctl cluster_status"
  # To parse the result string, this function normalize the output string
  # - Removing first line
  # - Removing "... Done" : old version returns this
  cmd = 'rabbitmqctl cluster_status'
  Chef::Log.debug("[rabbitmq_cluster] Executing #{cmd}")
  cmd = get_shellout(cmd)
  cmd.run_command
  cmd.error!
  result = cmd.stdout.split(/\n/, 2).last.squeeze(' ').gsub(/\n/, '').gsub('...done.', '')
  Chef::Log.debug("[rabbitmq_cluster] rabbitmqctl cluster_status : #{result}")
  result
end
```

Comment (jjasghar) [inline providers/cluster.rb]: :+1:

**Comment by sunggun-yu** [comment]
@jjasghar @cmluciano 
seems like,,,, Travis has been stopped by connection timed out...
https://travis-ci.org/jjasghar/rabbitmq/builds/53843704

**Comment by jjasghar** [comment]
I restarted it, lets see if it passes.

**Comment by sunggun-yu** [comment]
@jjasghar Looks good!

**Comment by jesseadams** [comment]
:+1:

File: recipes/cluster.rb [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+cluster_nodes = node['rabbitmq']['clustering']['cluster_nodes']
+cluster_nodes = cluster_nodes.to_json
+
+if node['rabbitmq']['cluster']
+  # Manual clustering
+  unless node['rabbitmq']['clustering']['use_auto_clustering']
+    # Join in cluster
+    rabbitmq_cluster cluster_nodes do
+      action :join
+    end
```
Comment (jesseadams) [inline recipes/cluster.rb]: Should this be in a `not_if {}` ?

Comment (sunggun-yu) [inline recipes/cluster.rb]: @jesseadams 
node['rabbitmq']['clustering']['use_auto_clustering'] is the flag for using auto clustering or manual clustering.  should I use `not_if`? can you please explain to me how `not_if` different from `unless` at this code?

Comment (jesseadams) [inline recipes/cluster.rb]: It is fundamentally the same. I thought that not using not_if was a FoodCritic violation in this case. Maybe not?

**Comment by jesseadams** [comment]
Can we please get this reviewed once more and merged? Thanks!
